### PR TITLE
`__inputname` instead of `typename`

### PR DESCRIPTION
**Pull request by emeryotopalik**

### CODE REVIEW

File: lib/absinthe/phase/schema.ex [hunk 1]
```diff
@@ -161,6 +161,9 @@ defmodule Absinthe.Phase.Schema do
 
   defp set_schema_node(%Blueprint.Input.Field{} = node, parent, schema, adapter) do
     case node.name do
+      name = "__inputname" ->
+        %{node | schema_node: find_schema_field(parent.schema_node, name, node, schema, adapter)}
+
```
Comment (emeryotopalik) [inline lib/absinthe/phase/schema.ex:166]: necessary

File: lib/absinthe/schema/rule/type_names_are_reserved.ex [hunk 2]
```diff
@@ -53,6 +53,10 @@ defmodule Absinthe.Schema.Rule.TypeNamesAreReserved do
     check_named(schema, type, "argument", arg)
   end
 
+  defp check_named(_schema, _type, "field", %{name: "__inputname"}) do
```
Comment (emeryotopalik) [inline lib/absinthe/schema/rule/type_names_are_reserved.ex:56]: necessary

File: lib/absinthe/phase/schema.ex [hunk 3]
```diff
@@ -237,6 +240,14 @@ defmodule Absinthe.Phase.Schema do
           Absinthe.Schema.t(),
           Absinthe.Adapter.t()
         ) :: nil | Type.Field.t()
+  defp find_schema_field(%{fields: fields}, "__inputname" = name, node, schema, adapter) do
+    internal_name = adapter.to_internal_name(name, :field)
+
+    fields
+    |> Map.values()
+    |> Enum.find(&match?(%{name: ^internal_name}, &1))
```
Comment (emeryotopalik) [inline lib/absinthe/phase/schema.ex]: necessary

and I prefer to keep the logic duplicated

File: lib/absinthe/phase/schema.ex (hunk 1 above)
Comment (binaryseed) [reply lib/absinthe/phase/schema.ex:166]: could you modify the `__` case to specifically allow this through

```
"__" <> introspection when introspection not in ["inputname"] ->
```

then we aren't duplicating the same code

File: lib/absinthe/phase/schema.ex (hunk 3 above)
Comment (binaryseed) [reply lib/absinthe/phase/schema.ex]: might be `:string` every time

File: lib/absinthe/schema/rule/input_union_input_name_present.ex [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  alias Absinthe.{Schema, Type}
+  require IEx
+
+  @moduledoc false
+
+  def explanation(%{
+        data: %{input_union_type: input_union_type, input_object_type: input_object_type}
+      }) do
+    """
+    The `__inputname` must be defined in a input_object :#{input_object_type} used in input_union :#{
```
Comment (binaryseed) [inline lib/absinthe/schema/rule/input_union_input_name_present.ex]: Rather than this validation, would you want to try and auto-inject the field ?

Comment (emeryotopalik) [reply lib/absinthe/schema/rule/input_union_input_name_present.ex]: Yah, I'll experiment with that

File: lib/absinthe/type/input_object.ex [hunk 5]
```diff
@@ -70,6 +70,7 @@ defmodule Absinthe.Type.InputObject do
     fields =
       attrs
       |> Keyword.get(:fields, [])
+      |> Kernel.++(__inputname: spliced_inputname())
```
Comment (binaryseed) [inline lib/absinthe/type/input_object.ex:73]: we haven't tackled the `introspection` queries at all yet, but we might want to make sure that this field does not show up in the list, but we do kinda want it to show up as something that can auto-complete.

we probably don't need to worry about that until we get a little more traction

Comment (emeryotopalik) [reply lib/absinthe/type/input_object.ex:73]: noted

**Comment by emeryotopalik** [comment]
@binaryseed Okay, this could use a quick review before being pushed to your branch. I updated the notes to track what we may still need to tackle/what is absolutely not clean about this.

**Comment by emeryotopalik** [comment]
@binaryseed okay. this works well enough for now. I add notes on the various, glaringly obvious shortcomings.

**Pull request merged**